### PR TITLE
feat: check the specified energy matches energy computed from nutrients

### LINKED ISSUE #2429: Quality facet - Nutrition - Compare stated and calculated energy for higher accuracy of product listings and spotting faulty products
### What
- Stated and calculated energy does not fit for approx. 14.000 products. Reason could be a simple type-o, a mistake in the label itself, but most often, it's editors that confuse kj and Kcal.

For example:
1. If calculated kcal is approx 4.2x higher than stated, then the editor probably meant KCal but wrote the energy in kj
2. If calculated kj is approx. 4.2x smaller than stated, then the editor probably meant kj but wrote the energy in KCal

### Proposed solution
An automatic test, something along ```Is kcal/100g approx. 9*fat+4*CH+4*protein (+/- 10%)```
If that test is not passed have it pass on to manual review, e.g. hunger games

These products would likely fail this test:
Sheet 1: Is calculated energy from fat+carbs+protein < 3800kj?--> 68 products
Sheet 2: Is calculated energy from fat+carbs+protein at least 80% of stated energy?--> 9000 products
Sheet 3: Is calculated energy from fat+carbs+protein max. 120% of stated energy? --> 5000 products
https://docs.google.com/spreadsheets/d/142UD8SrbOS2ZGNiaXLnNacw32Wtcia_iEMiPR2vxxAY/edit?usp=sharing

### Considered alternatives
1. Fix existing objects
2. Avoid future mistakes

Option 1: Calculate Kilocalories
Option 2: Calculate Kilojoules

Open question: Should g fibres be calculated with 7kj/g?

### Context
https://openfoodfacts.slack.com/archives/C03H290LF/p1569656651003000

### Part of
- #5538

### PR DESCRIPTION
**Pull request by stephanegigandet**
Fixes #2429 

This adds new data quality errors when the energy in kcal or in kj does not match the energy that can be computed from other nutrients.

### CODE REVIEW

File: lib/ProductOpener/DataQualityFood.pm [hunk 1]
```diff
@@ -512,6 +512,89 @@ sub check_carbon_footprint ($product_ref) {
...
+		return;
+	}
+
+	# Different countries allow different ways to determine energy
+	# One way is to compute energy from other nutrients
+	# We can thus try to use energy as a key to verify other nutrients
+
+	# See https://esha.com/blog/calorie-calculation-country/
+	# and https://eur-lex.europa.eu/legal-content/FR/TXT/HTML/?uri=CELEX:32011R1169&from=FR Appendix XIV
+
```
Comment (alexgarel) [inline lib/ProductOpener/DataQualityFood.pm:553]: Adding a 
```perl
	my %nutriments = $product_ref->{nutriments};
```
would make the code below more readable !

File: lib/ProductOpener/DataQualityFood.pm [hunk 2]
```diff
@@ -512,6 +512,89 @@ sub check_carbon_footprint ($product_ref) {
...
+			my $computed_energy = 0;
+			foreach my $nid (keys %{$energy_from_nutrients{europe}}) {
+
+				my $energy_per_gram = $energy_from_nutrients{europe}{$nid}{$unit};
+				my $grams = 0;
+				if ($nid =~ /_minus_/) {
+					my $nid_minus = $';
+					$nid = $`;
+					$grams -= $product_ref->{nutriments}{$nid_minus . "_value"} || 0;
+				}
```
Comment (alexgarel) [inline lib/ProductOpener/DataQualityFood.pm:575]:
```suggestion
				# handles nutriment1__minus__numtriment2 case
				if ($nid =~ /_minus_/) {
					my $nid_minus = $';
					$nid = $`;
					$grams -= $product_ref->{nutriments}{$nid_minus . "_value"} || 0;
				}
```

File: lib/ProductOpener/DataQualityFood.pm [hunk 3]
```diff
@@ -512,6 +512,89 @@ sub check_carbon_footprint ($product_ref) {
...
+				$grams += $product_ref->{nutriments}{$nid . "_value"} || 0;
+				$computed_energy += $grams * $energy_per_gram;
+			}
+
+			# Compare to specified energy value
+			if (   ($computed_energy < ($specified_energy * 0.9 - 5))
+				or ($computed_energy > ($specified_energy * 1.1 + 5)))
+			{
+				push @{$product_ref->{data_quality_errors_tags}},
+					"en:energy-value-in-$unit-does-not-match-value-computed-from-other-nutrients";
```
Comment (alexgarel) [inline lib/ProductOpener/DataQualityFood.pm:585]:
```suggestion
				# we have a quality problem
				push @{$product_ref->{data_quality_errors_tags}},
					"en:energy-value-in-$unit-does-not-match-value-computed-from-other-nutrients";
```

**Review by alexgarel — APPROVED** [review]
LGTM

**Review by CharlesNepote — APPROVED** [review]
Looks good to me.

**Pull request merged**